### PR TITLE
Support for sending PMs to email addresses

### PR DESCRIPTION
**Pull request by nbianca**

### CODE REVIEW

**Comment by SamSaffron** [comment]
Let us know when this is ready. 

Some stuff to consider:

- Send PM to email should be gated by trust level, out of the box only available to TL4 and above. 
- If there is no way to "reply" via email then the option should be disabled for staff
- Be sure to add some screenshots of how this works

**Comment by nbianca** [comment]
I fixed the merge conflicts. I have also tested and it works as expected.

**Comment by SamSaffron** [comment]
Looks like there are some rubocop failures ... can you install rubocop locally and run it on the files. 

Which editor are you using? 

@tgxworld we need a post on meta that explains the whole rubocop thing ASAP

File: lib/topic_creator.rb [hunk 1]
```diff
@@ -162,11 +162,20 @@ def process_private_message(topic)
...
 
-    unless @opts[:target_usernames].present? || @opts[:target_group_names].present?
+    unless @opts[:target_usernames].present? || @opts[:target_emails].present? || @opts[:target_group_names].present?
       rollback_with!(topic, :no_user_selected)
     end
 
+    if @opts[:target_emails].present?
+      if !SiteSetting.reply_by_email_enabled? then
+        rollback_with!(topic, :reply_by_email_disabled)
+      elsif @user.trust_level < 3 then
```
Comment (SamSaffron) [inline lib/topic_creator.rb]: This looks hardcoded to me, it really should be a site setting.

**Comment by SamSaffron** [comment]
we also need quite a few tests here  , especially for the server side / api side of things.

File: config/site_settings.yml [hunk 2]
```diff
@@ -830,6 +830,9 @@ trust:
   min_trust_to_send_messages:
     default: 1
     enum: 'TrustLevelSetting'
+  min_trust_to_send_email_messages:
+    default: 3
```
Comment (SamSaffron) [inline config/site_settings.yml]: I would default this to 4, 3 is too low.

Comment (SamSaffron) [inline config/site_settings.yml]: also, lets add one more setting so we can only enable this for staff. default is only staff are allowed this feature.

File: lib/topic_creator.rb [hunk 3]
```diff
@@ -213,4 +239,29 @@ def add_groups(topic, groups)
   def check_can_send_permission!(topic, obj)
     rollback_with!(topic, :cant_send_pm) unless @opts[:skip_validations] || @guardian.can_send_private_message?(obj)
   end
+
+  def find_or_create_user(email, display_name)
+    user = nil
+
+    User.transaction do
```
Comment (SamSaffron) [inline lib/topic_creator.rb]: Is there a reason for the transaction here? I should be wrapped in a parent transaction anyway

File: config/site_settings.yml (hunk 2 above)
Comment (nbianca) [inline config/site_settings.yml]: Ok, I will add that.

File: lib/topic_creator.rb (hunk 3 above)
Comment (nbianca) [inline lib/topic_creator.rb]: Makes sense. I missed the parent transaction.

File: config/locales/server.en.yml [hunk 4]
```diff
@@ -1010,6 +1011,7 @@ en:
     summary_max_results: "Maximum posts returned by 'Summary This Topic'"
 
     enable_private_messages: "Allow trust level 1 (configurable via min trust level to send messages) users to create messages and reply to messages. Note that staff can always send messages no matter what."
+    enable_private_email_messages: "Allow trust level 4 (configurable via min trust level to send messages) users to send private email messages. Note that staff can always send messages no matter what."
```
Comment (SamSaffron) [inline config/locales/server.en.yml]: the text here should be improved, this requires that enable_staged_users is true

**Comment by SamSaffron** [comment]
OK, lets try this out! thanks!

**Pull request merged**